### PR TITLE
Added a check for zero coefficient of variations for the Beta distribution

### PR DESCRIPTION
**Pull request by micheles**
Otherwise there would be a division by zero in the formulas
```python
    def _alpha(mean, stddev):
        return ((1 - mean) / stddev ** 2 - 1 / mean) * mean ** 2

    def _beta(mean, stddev):
        return ((1 - mean) / stddev ** 2 - 1 / mean) * (mean - mean ** 2)
```
This happened to Maria Camilla.

### CODE REVIEW

File: openquake/risklib/scientific.py [hunk 1]
```diff
@@ -131,6 +131,10 @@ def __init__(self, vf_id, imt, imls, mean_loss_ratios, covs=None,
                     pass
                 elif lr > 1:
                     raise ValueError('The meanLRs must be <= 1, got %s' % lr)
+                elif cov == 0:
```
Comment (raoanirudh) [inline openquake/risklib/scientific.py:134]: We also need to have a check prohibiting negative covs. This applies also for the 'LN' distribution.

For the `cov==0` case for both the 'BT' and 'LN' distributions, there are three subcases that must be handled differently:
1. All covs are zero, in which case we switch to the DegenerateDistribution. The situation in #6105 should be identical to this subcase.
2. `cov==0`, but also `lr==0`: this subcase is handled in the `if` statement just above.
3. `cov==0` with a non-negative `lr`: this is the subcase where we finally raise the `ValueError`.

Comment (raoanirudh) [reply openquake/risklib/scientific.py:134]: Also related to #4841, but at that time we only implemented the check on the upper bound.

File: openquake/risklib/scientific.py [hunk 2]
```diff
@@ -131,6 +131,10 @@ def __init__(self, vf_id, imt, imls, mean_loss_ratios, covs=None,
...
+                    raise ValueError(
+                        'Found a zero coefficient of variation in %s' %
+                        self.covs)
                 elif cov ** 2 > 1 / lr - 1:
                     # see https://github.com/gem/oq-engine/issues/4841
                     raise ValueError(
```
Comment (raoanirudh) [inline openquake/risklib/scientific.py:140]: We could improve the message to be more informative to something like: "The coefficient of variation %s > %s does not satisfy the requirement `0 < σ < sqrt[μ × (1 - μ)]` .."

Comment (raoanirudh) [reply openquake/risklib/scientific.py:140]: Meant to add the above comment to Lines 141–142 in the patched file.

File: openquake/risklib/scientific.py (hunk 1 above)
Comment (raoanirudh) [reply openquake/risklib/scientific.py:134]: Pointed out correctly by Maria Camila, there is also a 4th subcase for the 'BT' distribution:
`cov==0`, but `lr==1`: for this subcase, the beta distribution again becomes a one-point degenerate distribution with a Dirac delta function spike at the right end; the vulnerability function should always return a loss ratio of 1.0 in this subcase.

File: openquake/risklib/scientific.py (hunk 2 above)
Comment (raoanirudh) [reply openquake/risklib/scientific.py:144]: Message updated

File: openquake/risklib/scientific.py (hunk 1 above)
Comment (raoanirudh) [reply openquake/risklib/scientific.py:138]: All cases described above now handled, except the last one where `cov=0`, `lr=1` for the 'BT' distribution. This combination will be flagged as invalid. 

To correctly handle such cases, we must use a mixed discrete-continuous probability distribution, defined by a discrete probability mass assigned to `lr=0` and `lr=1`, and a continuous probability density for `lr∈(0, 1)`. Out of the scope of this PR.

**Review by raoanirudh — APPROVED** [review]

**Comment by micheles** [comment]
After many experiments, it turns out that special casing _alpha and _beta for mean=1 and stddev=0 does not work.
For mean=1 the alpha and beta coefficients are 1 and 0 respectively and the beta distribution is a delta function at x=1.
The issue is that if you do something like this
```python
res = numpy.zeros_like(alpha)
ok = beta !=0  # not singular
res[ok] = numpy.random.beta(alpha[ok], beta[ok])
res[~ok] = 1
```
this is not going to give results close to you want expect by setting stddev=.0000001 and mean=.00000001 (i.e. smooth while yo go through the limit) even if the the seed is fixed. The reason is that the random number generator will advantage differently.
Suppose the array size is 10 and there is a single singular value; the call to `numpy.random.beta(alpha, beta)` will advance the generator by 9 step, while if you regularize the singularity by using stddev=.0000001 and mean=.00000001 the random generator will advance by 10 step. The numbers produced by the beta distribution will be therefore quite different.
So I propose to keep things are they are now: stddev=0 is an error and the user has to regularize the situation manually, the engine will not perform any magic to replace the beta distribution with a delta distribution or something like that.

**Pull request merged**